packages/exhibitor/extra/dcos_zk_backup.py

```diff
@@ -0,0 +1,322 @@
+#!/usr/bin/env python3
+"""
+Command line tool for automated DC/OS ZooKeeper instance backup and restore.
+
+This script is made available as `dcos-zk` executable in DC/OS.
+
+"""
+import shlex
+import shutil
+import subprocess
+import sys
+import tarfile
+from argparse import ArgumentParser, ArgumentTypeError
+from pathlib import Path
+from tempfile import TemporaryDirectory
+
+
+EXHIBITOR_DIR = Path('/var/lib/dcos/exhibitor')
+
+
+def run_command(cmd: str, verbose: bool) -> None:
+    """
+    Run a command in a subprocess.
+
+    Args:
+        verbose: Show the output.
+
+    Raises:
+        subprocess.CalledProcessError: The given cmd exits with a non-0 exit code.
+    """
+    stdout = None if verbose else subprocess.PIPE
+    stderr = None if verbose else subprocess.STDOUT
+    subprocess.run(
+        args=shlex.split(cmd),
+        stdout=stdout,
+        stderr=stderr,
+        check=True,
+    )
+
+
+def _copy_dir_and_preserve_ownership(src: Path, dst: Path, verbose: bool) -> None:
+    """
+    Copy a directory from ``src`` to ``dst`` and preserve ownership.
+
+    We need to preserve ownership for a successful restore.
+    We use a subprocess (with `cp -p`) for this rather than Python's
+    `shutil.copytree` as `copytree` does not preserve ownership.
+    """
+    run_command('cp -prv {src} {dst}'.format(src=src, dst=dst), verbose)
+
+
+def backup_zookeeper(
+    backup: Path,
+    tmp_dir: Path,
+    verbose: bool,
+) -> None:
+    """
+    DC/OS ZooKeeper instance backup procedure.
+
+    ZooKeeper changes files while it is running.
+    In order to have a consistent backup stop ZooKeeper, run the backup procedure, then start ZooKeeper.
+
+    Running this procedure will temporarily bring down the ZooKeeper instance on this master node.
```

Comment:
This comment seems outdated. Running this function will not temporarily bring down the ZK instance.

I would prefer if we had a check (short circuit) that would check if the ZK instance is running and if is it would exit with non-0 exit code.

Reply:
Thanks, missed that.